Improve robustness against decoding errors when using V100 protocol.

mfrener opened the pull request:
While issue https://github.com/stoqey/ib/issues/23 is fixed, it showed two additional issues that are addressed by this PR:

1) The V100 allows to check for message boundaries. It sends the length of the message, followed by the tokens.
So we can verify if a decoder function actually did read all token it actually needs to read. Also we can drain any remaining any remaining tokens up next message start, so that decoding of other messages continues to to work, also if one decoder function is buggy.
Also we can verify if a decoder function tries to read past the message boundary and stop it form doing so.

2) There no useful error output at all if a decoding error happens - did improve that.
For testing purpose I did modify decodeMsg_POSITION to read one token than is it should. Decoding of all other messages does continue, but you will get an error event with:
Error: Underrun error on decodeMsg_POSITION: End of message reached. Please report to https://github.com/stoqey/ib
    at Controller.emitError (C:\srcmfr\ib\src\io\controller.ts:202:37)
    at Decoder.process (C:\srcmfr\ib\src\io\decoder.ts:285:25)
    at Controller.processIngressQueue (C:\srcmfr\ib\src\io\controller.ts:93:18)
    at Socket.onMessage (C:\srcmfr\ib\src\io\socket.ts:237:23)
    at Socket.onData (C:\srcmfr\ib\src\io\socket.ts:189:14)
    at Socket.<anonymous> (C:\srcmfr\ib\src\io\socket.ts:108:32)
    at Socket.emit (events.js:210:5)
    at addChunk (_stream_readable.js:309:12)
    at readableAddChunk (_stream_readable.js:290:11)
    at Socket.Readable.push (_stream_readable.js:224:10)
    at TCP.onStreamRead (internal/stream_base_commons.js:182:23)
If the functions reads less tokens than it should, decoding of all other messages also continues to work and the error event is:
Error: Decoding error on decodeMsg_POSITION: unprocessed data left on queue. Please report to https://github.com/stoqey/ib
    at Controller.emitError (C:\srcmfr\ib\src\io\controller.ts:202:37)
    at Decoder.process (C:\srcmfr\ib\src\io\decoder.ts:271:27)
    at Controller.processIngressQueue (C:\srcmfr\ib\src\io\controller.ts:93:18)
    at Socket.onMessage (C:\srcmfr\ib\src\io\socket.ts:237:23)
    at Socket.onData (C:\srcmfr\ib\src\io\socket.ts:189:14)
    at Socket.<anonymous> (C:\srcmfr\ib\src\io\socket.ts:108:32)

ajanian reviewed (approved):
This looks fine to me.  If things continue to diverge in V100 / non-V100 protocol we may want to think about refactoring this to an abstract class with the shared logic.  It isn't too bad to follow yet.

mfrener commented:
> This looks fine to me. If things continue to diverge in V100 / non-V100 protocol we may want to think about refactoring this to an abstract class with the shared logic. It isn't too bad to follow yet.

Yes agree - I actually has same in mind, but with different long term plan. :D 
Instead of doing an abstraction (i.e. replace string[] with a TokenStream class that handles V100 / non-V100 protocol transparently) I added two code paths. 
I think long-term we should drop non-V100 support - it will dead code, as at some point everything will have updated his stone age TWS / IB Gateway version to current version. At that point we simply delete the non-V100 code-path.

ajanian commented:
dropping support is even better.  I assumed you wanted to keep support for some reason

On Fri, Jan 29, 2021 at 8:14 AM, Matthias Frener < notifications@github.com > wrote:

> 
> 
>> 
>> 
>> This looks fine to me. If things continue to diverge in V100 / non-V100
>> protocol we may want to think about refactoring this to an abstract class
>> with the shared logic. It isn't too bad to follow yet.
>> 
>> 
> 
> 
> 
> Yes agree - I actually has same in mind, but with different long term
> plan. :D
> Instead of doing an abstraction (i.e. replace string[] with a TokenStream
> class that handles V100 / non-V100 protocol transparently) I added two
> code paths.
> I think long-term we should drop non-V100 support - it will dead code, as
> at some point everything will have updated his stone age TWS / IB Gateway
> version to current version. At that point we simply delete the non-V100
> code-path.
> 
> 
> 
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub (
> https://github.com/stoqey/ib/pull/27#issuecomment-769797873 ) , or unsubscribe
> (
> https://github.com/notifications/unsubscribe-auth/AAAYLBHMTNALUMEQDDMQ3BLS4KYFBANCNFSM4WXSA7RA
> ).
> 
> 
>

mfrener commented:
> dropping support is even better.  I assumed you wanted to keep support for some reason
> […](#)
> On Fri, Jan 29, 2021 at 8:14 AM, Matthias Frener < ***@***.*** > wrote: > > > This looks fine to me. If things continue to diverge in V100 / non-V100 > protocol we may want to think about refactoring this to an abstract class > with the shared logic. It isn't too bad to follow yet. > > Yes agree - I actually has same in mind, but with different long term plan. :D Instead of doing an abstraction (i.e. replace string[] with a TokenStream class that handles V100 / non-V100 protocol transparently) I added two code paths. I think long-term we should drop non-V100 support - it will dead code, as at some point everything will have updated his stone age TWS / IB Gateway version to current version. At that point we simply delete the non-V100 code-path. — You are receiving this because your review was requested. Reply to this email directly, view it on GitHub ( [#27 (comment)](https://github.com/stoqey/ib/pull/27#issuecomment-769797873) ) , or unsubscribe ( https://github.com/notifications/unsubscribe-auth/AAAYLBHMTNALUMEQDDMQ3BLS4KYFBANCNFSM4WXSA7RA ).



> dropping support is even better.  I assumed you wanted to keep support for some reason
> […](#)
> On Fri, Jan 29, 2021 at 8:14 AM, Matthias Frener < ***@***.*** > wrote: > > > This looks fine to me. If things continue to diverge in V100 / non-V100 > protocol we may want to think about refactoring this to an abstract class > with the shared logic. It isn't too bad to follow yet. > > Yes agree - I actually has same in mind, but with different long term plan. :D Instead of doing an abstraction (i.e. replace string[] with a TokenStream class that handles V100 / non-V100 protocol transparently) I added two code paths. I think long-term we should drop non-V100 support - it will dead code, as at some point everything will have updated his stone age TWS / IB Gateway version to current version. At that point we simply delete the non-V100 code-path. — You are receiving this because your review was requested. Reply to this email directly, view it on GitHub ( [#27 (comment)](https://github.com/stoqey/ib/pull/27#issuecomment-769797873) ) , or unsubscribe ( https://github.com/notifications/unsubscribe-auth/AAAYLBHMTNALUMEQDDMQ3BLS4KYFBANCNFSM4WXSA7RA ).

On twsapi@groups.io there are some posts from ppl still running old TWS versions from time to time (for whatever reason..). That is about the only reason we still have that code - if someone is forced to run an old Gateway. Lets wait some more time, maybe until next official API IB update  - pretty sure IB itself will also drop it at some point.  If they keep it, we remove disableV100 function form API and if nobody complains, we remove the internal IO code as well ;)
